indexers/vector/ZarrIndexer/__init__.py

```diff
@@ -0,0 +1,63 @@
+from os import path
+from typing import List, Union, Optional
+
+import zarr
+import numpy as np
+
+from jina.executors.indexers.vector import NumpyIndexer
+from jina.helper import cached_property
+
+
+class ZarrIndexer(NumpyIndexer):
+    """Indexing based on Zarr arrays
+    
+    For more information about Zarr, please check
+    https://zarr.readthedocs.io/en/stable/index.html
+    
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        
+    def get_add_handler(self) -> zarr.hierarchy.Group:
+        """Open an existing zarr group file for adding new vectors
+
+        :return: a zarr group file
+        """
+        return zarr.open(store=self.index_abspath, mode='a')
+
+    def get_create_handler(self) -> zarr.hierarchy.Group:
+        """Create a zarr group file for adding new vectors
+
+        :return: a zarr group file
+        """
+        return zarr.open(store=self.index_abspath, mode='w')
+    
+    def add(self, keys: 'np.ndarray', vectors: 'np.ndarray', *args, **kwargs) -> None:
+        self._validate_key_vector_shapes(keys, vectors)
+        if 'default' in self.write_handler.array_keys():
+            self.write_handler['default'].append(data=vectors)
+        else:
+            self.write_handler.array(name='default', data=vectors)
+        self.key_bytes += keys.tobytes()
+        self.key_dtype = keys.dtype.name
+        self._size += keys.shape[0]
+    
+    @property
+    def query_handler(self):
+        return self.get_query_handler()
+    
+    def get_query_handler(self) -> Optional['zarr.core.Array']:
+        if not (path.exists(self.index_abspath) or self.num_dim or self.dtype):
+            return
+
+        if hasattr(self, 'write_handler') and 'default' in self.write_handler.array_keys():
```

Comment:
it feels weird to depend on `write_handler` when querying, why is this option needed?

Reply:
This is just to check if the `default` array is present in the group that was created during `write_handler` creations. Instead, we can open the `zarr Group` first, then check for `default` array. That would be 2 `zarr.open` calls. This seems better.